### DISCUSSION TITLE
Problem with tensorflow.keras : No module named 'tensorflow.keras'

### DISCUSSION THREAD

**Discussion by vmielecnrs** [pyinstaller/pyinstaller#6519] [Help] [answered]
Hi,

I don't manage to make `PyInstaller` work with `TensorFlow-Keras` (see [this post](https://www.pyimagesearch.com/2019/10/21/keras-vs-tf-keras-whats-the-difference-in-tensorflow-2-0/) for details about the difference with `Keras`).

Let's consider this very simple script `test.py` :
```
import tensorflow as tf
from tensorflow.keras.layers import Dense,Flatten,Dropout
from tensorflow.keras.models import Sequential

model = Sequential([
  Flatten(input_shape=(28, 28)),
  Dense(128, activation='relu'),
  Dropout(0.2),
  Dense(10)
])
```
I use `PyInstaller` the classical way: ` pyinstaller --onedir test.py`

and finally execute `dist/test/test`, I got the error
```
  File "test.py", line 4, in <module>
    from tensorflow.keras.layers import Dense,Flatten,Dropout
ModuleNotFoundError: No module named 'tensorflow.keras'
```

This problem is not solved using `--collect-all tensorflow.keras`. I tried to implement different hooks, use hidden imports,  but also the same error.

**I think there is a problem with internal cross-links between TensorFlow and Keras librairies, cross-links that we don't see but are managed by Tensorflow ?**

_My environment: Linux Manjaro, Python 3.10, PyInstaller 4.8, Tensorflow 2.8.0-rc0_

---
Edit : I still have the problem with the modified version:

```
import tensorflow as tf
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])
```
I got 
```
File "test.py", line 5, in <module>
    model = tf.keras.models.Sequential([
  File "tensorflow/python/util/lazy_loader.py", line 58, in __getattr__
  File "tensorflow/python/util/lazy_loader.py", line 41, in _load
  File "importlib/__init__.py", line 126, in import_module
...
ModuleNotFoundError: No module named 'keras.api'
```

**Comment by rokm** [comment, marked as answer]
> ModuleNotFoundError: No module named 'keras.api'

How about a `--collect-submodules keras`?

**Comment by vmielecnrs** [comment]
Thanks @rokm  !

Using `--collect-submodules keras` solved the problem. I'm sorry I didn't think of it...
